### ISSUE TITLE
render markdown in github/reddit context

### ISSUE THREAD

**Issue by purarue** [karlicoss/promnesia#233]
Following up on from the discussion from #232 

~~Tried an initial attempt at trying to render the markdown for `my.discord` context using `mistletoe.Document.render` in the backend, but I don't think it wouldn't scale well - especially if you have a couple thousand github/reddit comments - would estimate it would take close to an 30 minutes just to index...~~

Edit: wasn't `mistletoe` causing the lag, see comment below

Perhaps a flag could be added to `github` like

```
def index(render_markdown: bool = False):
 ....
```

to optionally enable it?

---

The other option is to just tag it as markdown in the `promnesia index` phase with another hint like !markdown and check for that while rendering the tag in the extension.

On any singular page, the amount of reddit/github comments found by promnesia is relatively low (so typically wouldn't cause lots of lag?) compared to something like browser history, so this may be the better strategy -- would likely need a library to handle that in JS. Should probably have a option to enable/disable that as well

Not a high priority issue, may try to see what I can figure out once I'm a bit more familiar with building/developing on the browser extension here

**Comment by purarue** [comment]
Not sure if the backend strategy is viable, but ended up writing this helper to extend the `promnesia.sources.markdown` module:

```diff
diff --git a/src/promnesia/sources/markdown.py b/src/promnesia/sources/markdown.py
index af1d338..2b9bde2 100644
--- a/src/promnesia/sources/markdown.py
+++ b/src/promnesia/sources/markdown.py
@@ -99,3 +99,39 @@ def extract_from_file(fname: PathIsh) -> Iterator[Extraction]:
                 locator=Loc.file(fname), # TODO line number
                 context=r.context,
             )
+
+
+class TextParser(Parser):
+    '''
+    Used to extract links/render markdown from text, e.g. reddit/github comments
+    Instead of chunking blocks like for files, this returns the entire
+    message rendered as the context
+    '''
+    def __init__(self, text: str):
+        self.doc = mistletoe.Document(text)
+
+
+    def _doc_ashtml(self):
+        '''
+        cached html representation of the entire html message/document
+        '''
+        if not hasattr(self, '_html'):
+            self._html = HTML_MARKER + _ashtml(self.doc)
+        return self._html
+
+
+    def _extract(self, cur, last_block = None) -> Iterator[Parsed]:
+        if not isinstance(cur, (AutoLink, Link)):
+            return
+
+        yield Parsed(url=cur.target, context=self._doc_ashtml())
+
+
+def extract_from_text(text: str) -> Iterator[Result]:
+    '''
+    assume this is rendering something like a github/reddit markdown message
+    use the entire contents of the comment/body as the context
+    '''
+    # note: returns Result (link/context), not Visit
+    # the callee function has to insert dt/duration etc.
+    yield from TextParser(text).walk()
```

Not sure if its going to be used, so might as well leave it in this comment; Let me know if I should PR it

**Comment by seanbreckenridge** [comment]
Actually, seems like a lot of the lag remained after removing mistletoe, it may be URLExtract thats at fault:

![](https://i.imgur.com/48pAlSD.png)

**Comment by purarue** [comment]
Ah, a lot of the lag was more likely because I was calling the `URLExtract()` in a loop; every time I was extracting the link -- which was loading info from cache/doing extra work

![](https://i.imgur.com/48pAlSD.png`)

Will do another test to see if running `mistletoe` `render` on the backend during `promnesia index` for github/reddit is reasonable

**Comment by purarue** [comment]
Yeah, just to confirm; `mistletoe` is fine, will push a possible PR in a bit if rendering markdown as HTML on the backend is something you're fine with.

**Comment by karlicoss** [comment]
> Perhaps a flag could be added to github to optionally enable it?

Yep :+1:  was just about to suggest it, because one thing it might break is inline highlights, because the current algorithm is pretty trivial -- it ignores punctuation but won't strip out HTML tags. So might be nicer to make this behaviour opt-in for now.

Overall I feel like eventually the right solution will be either sending both HTML bit and the original text from the backend (could be even rendered to HTML by the server & cached), or actually render it in the browser (and here HTML could also be cached, in case we do experience performance issues).


> Ah, a lot of the lag was more likely because I was calling the URLExtract() in a loop; every time I was extracting the link

Yep, it's a bit slow to initialize! That's why the factory in common caches it https://github.com/karlicoss/promnesia/blob/08bd031967a1d8223efe65c126b7a2c7c438c12d/src/promnesia/common.py#L197-L206

> will push a possible PR in a bit if rendering markdown as HTML on the backend is something you're fine with.

Yeah -- definitely something I've been meaning to experiment with, so would be glad! 

P.S. oh, btw just remembered, in Orger I've been using `pandoc` before to convert some of markdown/html into org-mode: https://github.com/karlicoss/orger/blob/master/src/orger/pandoc.py 

Not sure what would be the difference with mistletoe, in theory pandoc is more mature (especially for markdown -> HTML), and in theory if we run it as a process (inspead of spawning process per every bit of markdown) and in parallel, it might be faster than mistletoe. So might be interesting to try it out, but also might not make any difference -- though just will leave it here.

**Comment by purarue** [comment]
On pandoc vs. mistletoe

Wrote a [little script to just do a visual comparison](https://github.com/seanbreckenridge/HPI/blob/d9e31d5008858b4249761e28b08408cff2d525e9/scripts/markdown_test)

Obvious benefit of mistletoe is that is remains in python, Pandoc has a couple edge cases that may be handled nicer but I can't see a obvious reason to switch

Pandoc tends to remove newlines, but that doesn't really matter when its rendered as markdown

It adds a couple extra semantic tags:

```
PANDOC:
<p><span class="citation" data-cites="seanbreckenridge">@seanbreckenridge</span> &gt; Could I ask what version of python/OS you’re using? &gt; &gt; Seems <code>getchildren</code> may be deprecated.</p><p>Python 3.9.0a3+ on Debian Experimental</p><blockquote><p>I pushed a <a href="https://github.com/seanbreckenridge/keepassxc-pwned/commit/dfbf23ae0508a44f6b38cd5cd757e032ed58f934">fix</a> to a different branch, could you try:</p><pre><code>pip3 uninstall -y keepassxc-pwned
pip3 install git+https://github.com/seanbreckenridge/keepassxc-pwned@fix-deprecation</code></pre><p>and run again?</p></blockquote><p>It worked! merge it. But at first it had some errors about <code>api.pwnedpasswords.com</code> not being resolved. It was horrible, I found my bank account’s password to be in the database.</p>
---
MISTLETOE:
<p>@seanbreckenridge</p>
<blockquote>
<p>Could I ask what version of python/OS you're using?</p>
<p>Seems <code>getchildren</code> may be deprecated.</p>
</blockquote>
<p>Python 3.9.0a3+ on Debian Experimental</p>
<blockquote>
<p>I pushed a <a href="https://github.com/seanbreckenridge/keepassxc-pwned/commit/dfbf23ae0508a44f6b38cd5cd757e032ed58f934">fix</a> to a different branch, could you try:</p>
<pre><code>pip3 uninstall -y keepassxc-pwned
pip3 install git+https://github.com/seanbreckenridge/keepassxc-pwned@fix-deprecation
</code></pre>
<p>and run again?</p>
</blockquote>
<p>It worked! merge it.
But at first it had some errors about <code>api.pwnedpasswords.com</code> not being resolved.
It was horrible, I found my bank account's password to be in the database.</p>
```

Replaces punctuation where it makes sense to:

(quote)

```
PANDOC:
<p>I’ve clarified the statement a little bit.</p>
---
MISTLETOE:
<p>I've clarified the statement a little bit.</p>
------
```

(ellipsis)

```
PANDOC:
<p>has been working properly for months, closing…</p>
---
MISTLETOE:
<p>has been working properly for months, closing...</p>
------
```

**Comment by karlicoss** [comment]
Thanks! Yeah agree kind of nicer to keep it in python if possible. Also still could speedup even while using mistletoe with batching + + multiprocessing if we want, just would require a little more work.

**Comment by purarue** [comment]
solved by #236 and #235
